soc/xtensa/intel_adsp/common/ipc.c

```diff
@@ -207,12 +218,123 @@ static int dt_init(const struct device *dev)
 	return intel_adsp_ipc_init(dev);
 }
 
+#ifdef CONFIG_PM_DEVICE
+
+void intel_adsp_ipc_set_resume_handler(const struct device *dev,
+	intel_adsp_ipc_resume_handler_t fn, void *arg)
+{
+	struct ipc_control_driver_api *api =
+		(struct ipc_control_driver_api *)dev->api;
+	struct intel_adsp_ipc_data *devdata = dev->data;
+	k_spinlock_key_t key = k_spin_lock(&devdata->lock);
+
+	api->resume_fn = fn;
+	api->resume_fn_args = arg;
+
+	k_spin_unlock(&devdata->lock, key);
+}
+
+void intel_adsp_ipc_set_suspend_handler(const struct device *dev,
+	intel_adsp_ipc_suspend_handler_t fn, void *arg)
+{
+	struct ipc_control_driver_api *api =
+		(struct ipc_control_driver_api *)dev->api;
+	struct intel_adsp_ipc_data *devdata = dev->data;
+	k_spinlock_key_t key = k_spin_lock(&devdata->lock);
+
+	api->suspend_fn = fn;
+	api->suspend_fn_args = arg;
+
+	k_spin_unlock(&devdata->lock, key);
+}
+
+/**
+ * @brief Manages IPC driver power state change.
+ *
+ * @param dev IPC device.
+ * @param action Power state to be changed to.
+ * @return int Returns 0 on success or optionaly error code from the
+ * registered ipc_power_control_api callbacks.
+ *
+ * @note PM lock is taken at the start of each power transition to prevent concurrent calls
+ * to @ref pm_device_action_run function. Changing Device power state during ISR is not supported.
+ * If IPC Device performs hardware operation and device is busy (what should not happen)
+ * function returns failure. It is API user responsibility to make sure we are not entering
+ * device power transition while device is busy.
+ */
+static int ipc_pm_action(const struct device *dev, enum pm_device_action action)
+{
+	if (k_is_in_isr()) {
+		return -ESRCH;
+	}
+
+	if (pm_device_is_busy(INTEL_ADSP_IPC_HOST_DEV)) {
+		return -EBUSY;
+	}
+
+	const struct ipc_control_driver_api *api =
+		(const struct ipc_control_driver_api *)dev->api;
+
+	/* Return as default value when given PM action is not supported */
+	int ret = -ENOTSUP;
```

Comment:
This is probably causing problems  .... if `PM_DEVICE_ACTION_SUSPEND` and `api->suspend_fn == NULL` this function will return `ENOTSUP` which is wrong.